pex/dist_metadata.py

```diff
@@ -291,8 +291,10 @@ def from_filename(cls, path):
         # characters and dashes.
         fname = _strip_sdist_path(path)
         if fname is not None:
-            project_name, version = fname.rsplit("-", 1)
-            return cls(project_name=project_name, version=version)
+            components = fname.rsplit("-", 1)
+            if len(components) == 2:
+                project_name, version = components
+                return cls(project_name=project_name, version=version)
```

Comment:
This is the wrong place to fix and, as you suggest, just not trying to handle non-wheels at all in the locker should work now.

Reply:
Ah, right - no. This is complicated. I'll need to revisit the case of additional artifacts. You can't get additional artifacts for an archive URL direct reference - direct reference requirements are always singular. But for a normal sdist from PyPI, it can be just one of many archives that serve as artifacts for a universal lock of a particular pin - and this is where the current importance of identifying a Pin lies. More work on this tomorrow am...

Reply:
Ok, this was indeed complicated, but really only due to organic growth in the Locker logic. I took a clean-slate approach to that and it's now much simpler. The upshot is that parsing name and version from an artifact is now only ever done for wheels - which it's guaranteed to work for, and for links Pip identified as candidates via the file name / index structure. These can include sdists that are never built and we can rely on Pip here and avoid a build just to get name / version it has already blessed.